icaruscode/CRT/CRTTPCTruthEffMCKalman_module.cc

```diff
@@ -0,0 +1,1721 @@
+	      reco::PrincipalComponents::EigenVectors recobEigenVecs = eigenMat.eigenvectors().cast<float>();
+
+	      std::cout << "Eigenvalues are: " << recobEigenVals(0) << ", " << recobEigenVals(1) << ", " << recobEigenVals(2) << std::endl;
+
+	      float maxeval = FLT_MIN; int maxvalpos = -1;
+
+	      for(int tempindex=0; tempindex<3; tempindex++){
+
+			if(maxeval<(float)recobEigenVals(tempindex)) {
+
+				maxeval = (float)recobEigenVals(tempindex);
+				maxvalpos = tempindex;
+
+			}//end if(maxeval<(float)recobEigenVals)
+
+	      }//end for(int tempindex=0; tempindex<3; tempindex++)
+
+	      bestfit_dir.SetXYZ(recobEigenVecs(0,maxvalpos), recobEigenVecs(1,maxvalpos), recobEigenVecs(2,maxvalpos));
+
+	}//end if(eigenMat.info() == Eigen::ComputationInfo::Success)
+
+	}//end if(!forcedbreak)
+	std::cout << "End endpoint_PCA_ana function............\n-----------------------------------------\n\n\n";
+
+}//end definition of void endpoint_PCA_ana(recob::Track trk, int startorend, TVector3 &bestfit_dir, TVector3 &bestfit_pos)
+
+void icarus::CRTTPCTruthEffMCKalman::pointChecker(recob::Track trk, bool &startisinTPC, bool &endisinTPC, int &startflag, int &endflag){
+
+	geo::Point_t startpoint = trk.Start();
+	geo::Point_t endpoint = trk.End();
+
+	float xmin = 61.94;    float xmax = 358.49;
+	float ymin = -181.86;  float ymax = 134.96;
+	float zmin = -894.951; float zmax = 894.951;
+
+	bool x_start_check, x_end_check, y_start_check, y_end_check, z_start_check, z_end_check;
+
+	x_start_check = std::abs(startpoint.X())>=xmin && std::abs(startpoint.X())<=xmax;
+	x_end_check = std::abs(endpoint.X())>=xmin && std::abs(endpoint.X())<=xmax;
+	y_start_check = startpoint.Y()>=ymin && startpoint.Y()<=ymax;
+	y_end_check = endpoint.Y()>=ymin && endpoint.Y()<=ymax;
+	z_start_check = startpoint.Z()>=zmin && startpoint.Z()<=zmax;
+	z_end_check = endpoint.Z()>=zmin && endpoint.Z()<=zmax;
+
+	startisinTPC = x_start_check && y_start_check && z_start_check;
+	endisinTPC = x_end_check && y_end_check && z_end_check;
+
+	startflag=-1; endflag=-1;
+
+	if(startisinTPC) startflag=0;
+	else if(!x_start_check && y_start_check && z_start_check) startflag=1;
+	else if(x_start_check && !y_start_check && z_start_check) startflag=2;
+	else if(x_start_check && y_start_check && !z_start_check) startflag=3;
+	else startflag=4;
+	if(endisinTPC) endflag=0;
+	else if(!x_end_check && y_end_check && z_end_check) endflag=1;
+	else if(x_end_check && !y_end_check && z_end_check) endflag=2;
+	else if(x_end_check && y_end_check && !z_end_check) endflag=3;
+	else endflag=4;
+
+}//end definition of void icarus::CRTTPCTruthEffMCKalman::pointChecker(recob::Track trk, bool &startisinTPC, bool &endisinTPC)
+
+double icarus::CRTTPCTruthEffMCKalman::getnewDCA(TVector3 CRTHitpos, TVector3 TPC_posvec, TVector3 TPC_dirvec){
+
+    double denominator = TPC_dirvec.Mag();
+    double numerator = (CRTHitpos - TPC_posvec).Cross(CRTHitpos - TPC_posvec - TPC_dirvec).Mag();
+    return numerator/denominator;
```

Comment:
I suggest checking that the denominator isn't zero. If it is, you can set it to some large number.